### PR TITLE
Fix isEffectivelySingleton

### PR DESCRIPTION
**Pull request by odersky**
As usual, OrTypes need to be excluded. a.type | b.type is not effectively a singleton. It seems to be an easy trap to fall into.

Follow-up to #20474

### CODE REVIEW

File: tests/neg/i20474.scala [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class A
+class B extends A
+
+def f(a: A, c: A) =
+  val b1: a.type = a
+  val b2: a.type & B = a.asInstanceOf[a.type & B]
+  val b3: c.type & A = c
+  val b4: a.type | c.type = c
```
[the remark this answers is not in the document]
Comment (odersky) [reply tests/neg/i20474.scala:8]: I don't think that will work. Have you tried it? I see no string reason to make it work.

[the remark this answers is not in the document]
Comment (noti0na1) [reply tests/neg/i20474.scala:8]: It works in 3.4.2: https://scastie.scala-lang.org/A6PehvQvRhW2sUPVVtHiuA 
I think the `atoms` is doing all these checks.

[the remark this answers is not in the document]
Comment (odersky) [reply tests/neg/i20474.scala:8]: Ah, yes, atoms would do it. Bit it's an extreme corner case. Not sure we want to make sure it works in the general case. Because in this case we'd also have to accept

```scala
val b6: (a.type & A) | a.type
val d6: b6.type = a // ok
```
and I am not sure atoms would help us here.

**Comment by odersky** [comment]
I propose we merge without further refinements, since it plugs a recent soundness hole.

**Review by noti0na1 — APPROVED** [review]
